File: java/Quicksort.java
import java.util.Arrays;

public class Quicksort {

    /**
     * Ordena um array de números utilizando o algoritmo Quicksort.
     * @param arr Array a ser ordenado in-place
     */
    public static void ordenar(double[] arr) {
        if (arr == null || arr.length <= 1) return;
        executarQuicksort(arr, 0, arr.length - 1);
    }

    private static void executarQuicksort(double[] arr, int low, int high) {
        if (low < high) {
            int pivoIndex = particionar(arr, low, high);
            executarQuicksort(arr, low, pivoIndex - 1);
            executarQuicksort(arr, pivoIndex + 1, high);
        }
    }

    private static int particionar(double[] arr, int low, int high) {
        double pivo = arr[high];
        int i = low - 1;

        for (int j = low; j < high; j++) {
            if (arr[j] <= pivo) {
                i++;
                double temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        double temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }

    public static void main(String[] args) {
        double[] desordenado = {34, 7, 23, 32, 5, 62, 1, 9, 5, 0, -3};
        System.out.println("--- Quicksort ---");
        System.out.println("Array Desordenado: " + Arrays.toString(desordenado));
        ordenar(desordenado);
        System.out.println("Array Ordenado:   " + Arrays.toString(desordenado));
    }
}

File: javascript/quicksort.js
/**
 * Ordena um array de números utilizando o algoritmo Quicksort.
 * @param {number[]} arr
 * @returns {number[]} Novo array ordenado
 */
function quicksort(arr) {
  if (!Array.isArray(arr)) {
    throw new TypeError("Esperado um array de números");
  }
  if (arr.length <= 1) {
    return [...arr];
  }

  const pivo = arr[Math.floor(arr.length / 2)];
  const menores = [];
  const iguais = [];
  const maiores = [];

  for (const item of arr) {
    if (item < pivo) {
      menores.push(item);
    } else if (item === pivo) {
      iguais.push(item);
    } else {
      maiores.push(item);
    }
  }

  return [...quicksort(menores), ...iguais, ...quicksort(maiores)];
}

// Testes de exemplo
const desordenado = [34, 7, 23, 32, 5, 62, 1, 9, 5, 0, -3];
console.log("--- Quicksort ---");
console.log("Array Desordenado:", desordenado);
console.log("Array Ordenado:   ", quicksort(desordenado));

module.exports = { quicksort };
